components/exercises/chapter-4-oscar/Utils.tsx

```diff
@@ -70,3 +72,139 @@ export function reverseArrayInPlace(arr: any[]): undefined {
     // After the first iteration, the array becomes [5, 2, 3, 4, 1]
   }
 }
+
+/**
+ * Represents the properties of a list item.
+ */
+export type ListItemProp = {
+  value: ReactNode;
+  rest: ListItemProp | null;
+};
+
+/**
+ * Converts an array of numbers into a linked list.
+ * @param arr - The array of numbers to be converted.
+ * @returns The resulting linked list.
+ */
+export function arrayToList(arr: ReactNode[]): ListItemProp | null {
+  let list = null;
+
+  // arr.length - 1 --> because we want to start from the last element of the array as starts from 0
+  for (let i = arr.length - 1; i >= 0; i--) {
+    // list will store the result of every iteration, which starts on {value: 20, rest: null} so on
+    list = { value: arr[i], rest: list };
+  }
+  return list;
+}
+
+/**
+ * Converts a linked list to an array.
+ * @param list - The linked list to convert.
+ * @returns An array containing the values of the linked list.
+ */
+export function listToArray(list: ListItemProp | null): ReactNode[] {
+  let arr: ReactNode[] = [];
+
+  // node as condition because we want to stop when node is null
+  // node.rest because we want to move to the next node
+  for (let node: ListItemProp | null = list; node; node = node.rest) {
+    arr.push(node.value);
+  }
+  return arr;
+}
+
+/**
+ * Prepends a value to a list.
+ * @param value - The value to prepend to the list.
+ * @param list - The list to which the value will be prepended.
+ * @returns A new list with the value prepended to it.
+ */
+export function prepend(
+  value: number | string,
+  list: ListItemProp | null
+): ListItemProp | null {
+  // list is null when we reach the end of the list
+  return { value, rest: list };
+}
+
+/**
+ * Returns the value of the nth element in a list.
+ * @param list - The list to search for the nth element.
+ * @param n - The index of the element to retrieve.
+ * @returns The value of the nth element in the list, or undefined if the list is empty or n is out of bounds.
+ */
+export function nth(
+  list: ListItemProp | null,
+  n: number
+): undefined | ReactNode {
+  if (!list) return undefined;
+  else if (n == 0) return list.value;
```

Comment:
Better to use === instead of == https://www.phind.com/search?cache=jdlpiyfl2jwy64wjvo1rf1fo

Reply:
understood refactored